backends/ebpf/codeGen.cpp

```diff
@@ -458,6 +474,141 @@ void CodeGenInspector::widthCheck(const IR::Node *node) const {
             node, tb->size);
 }
 
+void CodeGenInspector::emitAndConvertByteOrder(const IR::Expression *expr, cstring byte_order) {
+    auto ftype = typeMap->getType(expr);
+    auto et = EBPFTypeFactory::instance->create(ftype);
+    unsigned widthToEmit = dynamic_cast<IHasWidth *>(et)->widthInBits();
+    cstring emit = "";
+    if (widthToEmit <= 16) {
+        emit = byte_order == "HOST" ? "bpf_ntohs" : "bpf_htons";
+    } else if (widthToEmit <= 32) {
+        emit = byte_order == "HOST" ? "bpf_ntohl" : "bpf_htonl";
+    } else if (widthToEmit <= 64) {
+        emit = byte_order == "HOST" ? "ntohll" : "bpf_cpu_to_be64";
+    }
+    builder->appendFormat("%s(", emit);
+    visit(expr);
+    builder->append(")");
+}
+
+void CodeGenInspector::emitTCBinaryOperation(const IR::Operation_Binary *b, bool isScalar) {
+    if (b->is<IR::Operation_Binary>()) {
+        b = b->to<IR::Operation_Binary>();
+    }
+    const IR::Expression *lexpr = b->left;
+    const IR::Expression *rexpr = b->right;
+    cstring stringop = b->getStringOp();
+
+    auto action = findContext<IR::P4Action>();
+    auto tcTarget = dynamic_cast<const P4TCTarget *>(builder->target);
+    cstring lByteOrder = "HOST", rByteOrder = "HOST";
+    if (lexpr) {
+        lByteOrder = tcTarget->getByteOrder(typeMap, action, lexpr);
+    }
+    if (rexpr) {
+        rByteOrder = tcTarget->getByteOrder(typeMap, action, rexpr);
+    }
+    if (lByteOrder == rByteOrder) {
+        visit(lexpr);
+        if (isScalar) {
+            builder->spc();
+            builder->append(stringop);
+            builder->spc();
+        } else {
+            builder->append(", &");
+        }
+        expressionPrecedence = b->getPrecedence() + 1;
+        visit(rexpr);
+        return;
+    }
+    if (lByteOrder == "NETWORK") {
+        // ConvertLeft
+        auto ftype = typeMap->getType(lexpr);
+        auto et = EBPFTypeFactory::instance->create(ftype);
+        unsigned width = dynamic_cast<IHasWidth *>(et)->widthInBits();
+        if (width <= 8) {
+            visit(lexpr);
+        } else {
+            emitAndConvertByteOrder(lexpr, "NETWORK");
+        }
+        if (isScalar) {
+            builder->spc();
+            builder->append(stringop);
+            builder->spc();
+        } else {
+            builder->append(", &");
+        }
+        expressionPrecedence = b->getPrecedence() + 1;
+        visit(rexpr);
+        return;
+    } else if (rByteOrder == "NETWORK") {
+        // ConvertRight
+        auto ftype = typeMap->getType(rexpr);
+        auto et = EBPFTypeFactory::instance->create(ftype);
+        unsigned width = dynamic_cast<IHasWidth *>(et)->widthInBits();
+        visit(lexpr);
+        if (isScalar) {
+            builder->spc();
+            builder->append(stringop);
+            builder->spc();
+        } else {
+            builder->append(", &");
+        }
+        expressionPrecedence = b->getPrecedence() + 1;
```

Comment:
What does this statement do? expressionPrecedence looks like global, would it impact ebpf backend in any way?